src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1956,6 +1968,570 @@ public Task OpenAsync(CancellationToken token)
+                        }
+                    }
+
+                    if (setInputSessionCloseHandle)
+                    {
+                       _inputSessionCloseHandle.Set();
+                    }
+                    if (cleanupContext)
+                    {
+                        requestContext.CloseAsync();
+                        cleanupContext = false;
+                    }
+                }
+                finally
+                {
+                    message.Close();
+                    if (cleanupContext)
+                    {
+                        requestContext.Abort();
+                    }
+                }
+            }
+
+            protected override void OnCloseResponseMessageReceived(RequestContext requestContext, Message message, SecurityProtocolCorrelationState correlationState, TimeSpan timeout)
+            {
+                bool cleanupContext = true;
+                try
+                {
+                    bool isCloseResponseExpected = false;
+                    bool setInputSessionCloseHandle = false;
+                    lock (LocalLock)
+                    {
+                        isCloseResponseExpected = _sentClose;
+                        if (isCloseResponseExpected && !_isInputClosed)
+                        {
+                            _isInputClosed = true;
+                            setInputSessionCloseHandle = true;
+                        }
+                    }
+                    if (!isCloseResponseExpected)
+                    {
+                        _duplexCommObj.Fault(new ProtocolException(SR.Format(SR.UnexpectedSecuritySessionCloseResponse)));
+                        return;
+                    }
+                    if (setInputSessionCloseHandle)
+                    {
+                        _inputSessionCloseHandle.Set();
+                    }
+
+                    requestContext.CloseAsync();
+                    cleanupContext = false;
+                }
+                finally
+                {
+                    message.Close();
+                    if (cleanupContext)
+                    {
+                        requestContext.Abort();
+                    }
+                }
+            }
+
+            internal bool WaitForOutputSessionClose(TimeSpan timeout, out bool wasAborted)
+            {
+                wasAborted = false;
+                try
+                {
+                    Task<bool> waitTask = _outputSessionCloseHandle.WaitAsync(timeout, false);
+                    return waitTask.GetAwaiter().GetResult();
+                }
+                catch (CommunicationObjectAbortedException)
+                {
+                    if (State != CommunicationState.Closed) throw;
+                    wasAborted = true;
+                    return true;
+                }
+            }
+
+            private bool WaitForInputSessionClose(TimeSpan timeout, out bool wasAborted)
+            {
```

Comment:
Same idea as in above method, make it async and move the out param to be put of a tuple return value.